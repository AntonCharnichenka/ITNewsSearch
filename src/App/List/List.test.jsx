import  { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent, within } from '@testing-library/react';
import { Item, List } from './index';

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};

const stories = [storyOne, storyTwo];

describe(
    'Item',
    () => {
        it(
            'renders all properties',
            () => {
                render(<Item item={storyOne}/>); // renders the element in jsdom environment
                // screen.debug(); // prints the html of the rendered element
                
                expect(screen.getByText('React')).toHaveAttribute('href', 'https://reactjs.org/'); // expect needs extesion by toHaveAttribute assertion method (in setup.js) 
                expect(screen.getByText('Jordan Walke')).toBeInTheDocument(); // same as for toHaveAttribute

            }
        );
        it(
            'renders a clickable dismiss button',
            () => {
                render(<Item item={storyOne}/>);
                expect(screen.getByRole('button')).toBeInTheDocument();
            }
        );
        it(
            'clicking dismiss button calls callback handler',
            () => {
                const handleRemoveItem = vi.fn();
                render(<Item item={storyOne} onRemoveItem={handleRemoveItem}/>)
                
                fireEvent.click(screen.getByRole('button'));

                expect(handleRemoveItem).toHaveBeenCalledTimes(1);
                expect(handleRemoveItem).toHaveBeenCalledWith(storyOne);
            }
        )
        it(
            'renders snapshot',
            () => {
                const { container } = render(<Item item={storyOne}/>);
                expect(container.firstChild).toMatchSnapshot();
            }
        )

    }
);

describe(
    'List',
    () => {
        it(
            'renders items and sort buttons', 
            () => {
                const onRemoveItem = vi.fn();
                render(<List list={stories} onRemoveItem={onRemoveItem}/>);

                expect(screen.getAllByRole('listitem')).toHaveLength(3);

                const sort_buttons = screen.getByText('Comments').closest('li');
                expect(within(sort_buttons).getAllByRole('button')).toHaveLength(4);
                expect(within(sort_buttons).getByText('Title')).toBeInTheDocument();
                expect(within(sort_buttons).getByText('Author')).toBeInTheDocument();
                expect(within(sort_buttons).getByText('Points')).toBeInTheDocument();

                const react_item = screen.getByText('React').closest('li');
                const redux_item = screen.getByText('Redux').closest('li');
                expect(within(react_item).getByText('Jordan Walke')).toBeInTheDocument();
                expect(within(redux_item).getByText('Dan Abramov, Andrew Clark')).toBeInTheDocument();
                const react_item_button = within(react_item).getByRole('button');
                expect(within(redux_item).getByRole('button')).toBeInTheDocument();

                fireEvent.click(react_item_button);
                expect(onRemoveItem).toHaveBeenCalledTimes(1);
                expect(onRemoveItem).toHaveBeenCalledWith(storyOne);
            }
        )
        it(
            'renders snapshot',
            () => {
                const { container } = render(<List list={stories}/>);
                expect(container.firstChild).toMatchSnapshot();
            }
        )
        it.todo('test sorting items')
    }
);
