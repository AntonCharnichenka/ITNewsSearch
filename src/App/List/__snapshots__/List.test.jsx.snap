// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item > renders snapshot 1`] = `
<li
  class="sc-beySPh cLibeB"
>
  <span
    class="sc-guDLey gQkZXT"
    width="40%"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    class="sc-guDLey gQjOKI"
    width="25%"
  >
    Jordan Walke
  </span>
  <span
    class="sc-guDLey gQjNCr"
    width="15%"
  >
    3
  </span>
  <span
    class="sc-guDLey gQlbYC"
    width="10%"
  >
    4
  </span>
  <span
    class="sc-guDLey gQlbYC"
    width="10%"
  >
    <button
      class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          id="Circle_Remove"
        >
          <g>
            <path
              d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
            />
            <path
              d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
            />
          </g>
        </g>
      </svg>
    </button>
  </span>
</li>
`;

exports[`List > renders snapshot 1`] = `
<ul>
  <li
    class="sc-beySPh cLibeB"
  >
    <span
      class="sc-guDLey gQkZXT"
      width="40%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        Title
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.5 17.25a.75.75 0 0 1-1.5 0V7.56l-2.22 2.22a.75.75 0 1 1-1.06-1.06l3.5-3.5a.75.75 0 0 1 1.06 0l3.5 3.5a.75.75 0 0 1-1.06 1.06L18.5 7.56v9.69Zm-15.75.25a.75.75 0 0 1 0-1.5h9.5a.75.75 0 0 1 0 1.5h-9.5Zm0-5a.75.75 0 0 1 0-1.5h5.5a.75.75 0 0 1 0 1.5h-5.5Zm0-5a.75.75 0 0 1 0-1.5h3.5a.75.75 0 0 1 0 1.5h-3.5Z"
          />
        </svg>
      </button>
    </span>
    <span
      class="sc-guDLey gQjOKI"
      width="25%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        Author
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.5 17.25a.75.75 0 0 1-1.5 0V7.56l-2.22 2.22a.75.75 0 1 1-1.06-1.06l3.5-3.5a.75.75 0 0 1 1.06 0l3.5 3.5a.75.75 0 0 1-1.06 1.06L18.5 7.56v9.69Zm-15.75.25a.75.75 0 0 1 0-1.5h9.5a.75.75 0 0 1 0 1.5h-9.5Zm0-5a.75.75 0 0 1 0-1.5h5.5a.75.75 0 0 1 0 1.5h-5.5Zm0-5a.75.75 0 0 1 0-1.5h3.5a.75.75 0 0 1 0 1.5h-3.5Z"
          />
        </svg>
      </button>
    </span>
    <span
      class="sc-guDLey gQjNCr"
      width="15%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        Comments
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.5 17.25a.75.75 0 0 1-1.5 0V7.56l-2.22 2.22a.75.75 0 1 1-1.06-1.06l3.5-3.5a.75.75 0 0 1 1.06 0l3.5 3.5a.75.75 0 0 1-1.06 1.06L18.5 7.56v9.69Zm-15.75.25a.75.75 0 0 1 0-1.5h9.5a.75.75 0 0 1 0 1.5h-9.5Zm0-5a.75.75 0 0 1 0-1.5h5.5a.75.75 0 0 1 0 1.5h-5.5Zm0-5a.75.75 0 0 1 0-1.5h3.5a.75.75 0 0 1 0 1.5h-3.5Z"
          />
        </svg>
      </button>
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        Points
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M18.5 17.25a.75.75 0 0 1-1.5 0V7.56l-2.22 2.22a.75.75 0 1 1-1.06-1.06l3.5-3.5a.75.75 0 0 1 1.06 0l3.5 3.5a.75.75 0 0 1-1.06 1.06L18.5 7.56v9.69Zm-15.75.25a.75.75 0 0 1 0-1.5h9.5a.75.75 0 0 1 0 1.5h-9.5Zm0-5a.75.75 0 0 1 0-1.5h5.5a.75.75 0 0 1 0 1.5h-5.5Zm0-5a.75.75 0 0 1 0-1.5h3.5a.75.75 0 0 1 0 1.5h-3.5Z"
          />
        </svg>
      </button>
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    />
  </li>
  <li
    class="sc-beySPh cLibeB"
  >
    <span
      class="sc-guDLey gQkZXT"
      width="40%"
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      class="sc-guDLey gQjOKI"
      width="25%"
    >
      Jordan Walke
    </span>
    <span
      class="sc-guDLey gQjNCr"
      width="15%"
    >
      3
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    >
      4
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Circle_Remove"
          >
            <g>
              <path
                d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
              />
              <path
                d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
              />
            </g>
          </g>
        </svg>
      </button>
    </span>
  </li>
  <li
    class="sc-beySPh cLibeB"
  >
    <span
      class="sc-guDLey gQkZXT"
      width="40%"
    >
      <a
        href="https://redux.js.org/"
      >
        Redux
      </a>
    </span>
    <span
      class="sc-guDLey gQjOKI"
      width="25%"
    >
      Dan Abramov, Andrew Clark
    </span>
    <span
      class="sc-guDLey gQjNCr"
      width="15%"
    >
      2
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    >
      5
    </span>
    <span
      class="sc-guDLey gQlbYC"
      width="10%"
    >
      <button
        class="sc-dmyCSP sc-hLQSwg iFoXrO fzPaEe"
        type="button"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <g
            id="Circle_Remove"
          >
            <g>
              <path
                d="M9.525,13.765a.5.5,0,0,0,.71.71c.59-.59,1.175-1.18,1.765-1.76l1.765,1.76a.5.5,0,0,0,.71-.71c-.59-.58-1.18-1.175-1.76-1.765.41-.42.82-.825,1.23-1.235.18-.18.35-.36.53-.53a.5.5,0,0,0-.71-.71L12,11.293,10.235,9.525a.5.5,0,0,0-.71.71L11.293,12Z"
              />
              <path
                d="M12,21.933A9.933,9.933,0,1,1,21.934,12,9.945,9.945,0,0,1,12,21.933ZM12,3.067A8.933,8.933,0,1,0,20.934,12,8.944,8.944,0,0,0,12,3.067Z"
              />
            </g>
          </g>
        </svg>
      </button>
    </span>
  </li>
</ul>
`;
